File: cracking_the_coding_interview_6th/chapter16/cpp/16_26_v01.cpp
#include <fstream>
#include <functional>
#include <string>
using namespace std;
using ll = long long;

double _eval(const string expr, ll head, ll last) {
  ll i = last;
  auto f = [&expr, &i, &head, &last](function<double(double, double)> op) {
    return op(_eval(expr, head, i - 1), _eval(expr, i + 1, last));
  };
  for (; i >= head; --i) {
    if (expr[i] == '+') return f(plus<double>());
    if (expr[i] == '-') return f(minus<double>());
  }
  i = last;
  for (; i >= head; --i) {
    if (expr[i] == '*') return f(multiplies<double>());
    if (expr[i] == '/') return f(divides<double>());
  }
  return stod(expr.substr(head, last - head + 1));
}

double eval(const string expr) {
  return _eval(expr, 0, expr.size() - 1);
}

void solve(const string &fp) {
  printf("# %s\n", fp.c_str());
  ifstream ifs{fp};
  string expr;
  ifs >> expr;
  printf("%f\n", eval(expr));
  puts("");
}

int main() {
  solve("../testcases/16_26/01.txt");
  solve("../testcases/16_26/02.txt");
  solve("../testcases/16_26/03.txt");
}

// # ../testcases/16_26/01.txt
// 23.500000

// # ../testcases/16_26/02.txt
// 120.000000

// # ../testcases/16_26/03.txt
// 99.000000

File: cracking_the_coding_interview_6th/chapter16/python/16_26_v01.py
from operator import add, sub, mul, truediv
from typing import Callable


def _eval(expr: str, head: int, last: int) -> float:
    def f(op: Callable[[float, float], float], i: int) -> float:
        return op(_eval(expr, head, i - 1), _eval(expr, i + 1, last))
    for i in range(last, head - 1, -1):
        if expr[i] == '+':
            return f(add, i)
        if expr[i] == '-':
            return f(sub, i)
    for i in range(last, head - 1, -1):
        if expr[i] == '*':
            return f(mul, i)
        if expr[i] == '/':
            return f(truediv, i)
    return int(expr[head:last + 1])


def eval(expr: str) -> float:
    return _eval(expr, 0, len(expr) - 1)


def ns(f):
    return next(f).strip()


def solve(fp: str):
    print(f"# {fp}")
    with open(fp) as f:
        expr = ns(f)
    print(eval(expr))
    print()


solve("../testcases/16_26/01.txt")
solve("../testcases/16_26/02.txt")
solve("../testcases/16_26/03.txt")

# # ../testcases/16_26/01.txt
# 23.5

# # ../testcases/16_26/02.txt
# 120

# # ../testcases/16_26/03.txt
# 99.0
